### ISSUE TITLE
What defines "appropriate" use of list comprehensions?

### ISSUE THREAD

**Issue by scouten** [doomspork/elixir-school#594]
The section http://elixirschool.com/lessons/basics/comprehensions/#basics ends with the sentence, "List comprehensions are syntactic sugar and should be used only when appropriate."

How do I know whether a possible use of a list comprehension is "appropriate"? If it isn't, what else would I use?

**Comment by holsky** [comment]
Coming from Python, aren't list comprehensions the way to go for shoveling data into, out of and between lists?

**Comment by ybur-yug** [comment]
I was pondering about this today and ended up browsing some source and messing around a bit @holsky @scouten.

```elixir
def process_foos(foos) do
  foos
  |> Enum.map(fn(foo) -> foo |> process_foo end)
end

def process_foo(foo) do
  # do some stuff to a foo
end
```

```elixir
def process_foos(foos), do: for foo <- foos, do: foo |> process_foo

def process_foo(foo) do
  # do some stuff to a foo
end
```

Lets use this simple bit as an example.
If we look at the [source where comprehensions are defined](https://github.com/elixir-lang/elixir/blob/50293b46f13a86328f0ffabdcbb8592e29ac24c6/lib/elixir/lib/kernel/special_forms.ex#L1228):

```elixir
  A comprehension accepts many generators and filters. Enumerable
  generators are defined using `<-`:
      # A list generator:
      iex> for n <- [1, 2, 3, 4], do: n * 2
      [2, 4, 6, 8]
      # A comprehension with two generators
      iex> for x <- [1, 2], y <- [2, 3], do: x*y
      [2, 3, 4, 6]
  Filters can also be given:
      # A comprehension with a generator and a filter
      iex> for n <- [1, 2, 3, 4, 5, 6], rem(n, 2) == 0, do: n
      [2, 4, 6]
  Note generators can also be used to filter as it removes any value
. . .
```

In the source docs, it really mentions nothing about a reason to use these sparingly.

If we peek at the actual source of the macro that does this [here](https://github.com/elixir-lang/elixir/blob/50293b46f13a86328f0ffabdcbb8592e29ac24c6/lib/elixir/lib/kernel/special_forms.ex#L1286) we can see that it really is just a simple piece and there are no warnings to be had in the source.

So I guess IMO this makes it come down to readability.

Is 

```elixir
  foos
  |> Enum.map(fn(foo) -> foo |> process_foo end)
```

better than

```elixir
def process_foos(foos), do: for foo <- foos, do: foo |> process_foo
```

I personally prefer the former. But I think this is a decision that ends up being something that can easily be taken at a team-level, as to my knowledge there is no official opinion on it in the styleguides the community has, either.

Thoughts, anyone?

**Comment by holsky** [comment]
I found this feeling "imperative" for the specific  case you do above, where you only apply a function to all elements in the pipe. I would write it like so, to make the flow more visible:

```elixir
foos 
  |> Enum.map(&process_foo/1)
```

In the examples from the documentation however, I would prefer the comprehension, because it actually creates a new list. So I agree, no reason to not use them :)

**Comment by ybur-yug** [comment]
@holsly yeah I totally should have used the `&stuff/arity` notation, youre right. Thats one of my 'i should do this but dont' problems in Elixir I need to work on.

**Comment by holsky** [comment]
Haha I'm glad I could contribute that :)
